From 50c1bf042b2a039dcc0477b77164025fd11c01fb Mon Sep 17 00:00:00 2001
From: Michael Himing <mhiming@gmail.com>
Date: Sun, 16 Dec 2018 13:07:33 +1100
Subject: [PATCH] Fix PlayerEditBookEvent

- Updating book writing (not signing) mutated the original item, making
it impossible to properly cancel the event or modify the book meta

- When the event was cancelled, the client's book would keep the
cancelled writing

diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 5c041c48fa..f63943f29f 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -823,10 +823,13 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
 
                             this.player.setSlot(enumitemslot, CraftEventFactory.handleEditBookEvent(player, enumitemslot, itemstack1, itemstack2)); // CraftBukkit
                         } else {
-                            ItemStack old = itemstack1.cloneItemStack(); // CraftBukkit
-                            itemstack1.a("pages", (NBTBase) itemstack.getTag().getList("pages", 8));
-                            CraftEventFactory.handleEditBookEvent(player, enumitemslot, old, itemstack1); // CraftBukkit
+                            // Paper start - dont mutate players current item, set it from the event
+                            ItemStack newBook = itemstack1.cloneItemStack();
+                            newBook.getOrCreateTagAndSet("pages", (NBTBase) itemstack.getTag().getList("pages", 8));
+                            this.player.setSlot(enumitemslot, CraftEventFactory.handleEditBookEvent(player, enumitemslot, itemstack1, newBook));
+                            // Paper end
                         }
+                        player.getBukkitEntity().updateInventory(); // Paper - fix client desync when event is cancelled
                     }
 
                 }
-- 
2.21.0

